### PR TITLE
Do not read HD entries from SQL schema resolver

### PR DESCRIPTION
**Pull request by devozerov**
This PR fixes a JVM crash that could occur when attempting to resolve SQL schema and there is an HD map.

Previously we read key-value samples from record stores for all maps. This is illegal for HD maps because they could be accessed only from partition threads.

This PR ensures that only concurrent HD indexes are used to read HD map samples.

Fixes: https://github.com/hazelcast/hazelcast/issues/17413

EE: https://github.com/hazelcast/hazelcast-enterprise/pull/3734

### CODE REVIEW

**Comment by devozerov** [comment]
Trst

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/PartitionedMapTableResolver.java [hunk 1]
```diff
@@ -174,11 +159,120 @@ private PartitionedMapTable createTable(
...
+    private static PartitionedMapTable emptyError(String mapName) {
         QueryException error = QueryException.error(
             "Cannot resolve IMap schema because it doesn't have entries on the local member: " + mapName
         );
 
         return new PartitionedMapTable(mapName, error);
     }
+
+    private static PartitionedMapTable hdError(String mapName) {
+        QueryException error = QueryException.error("Cannot query the IMap \"" + mapName
```
Comment (petrpleshachkov) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/PartitionedMapTableResolver.java:222]: We already have a similar exception trying to run a map scan without index. Shall we reuse the code?

Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/PartitionedMapTableResolver.java:222]: The exception you mentioned is moved here

Comment (petrpleshachkov) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/PartitionedMapTableResolver.java:222]: Ahh, I see that the code has been moved.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/PartitionedMapTableResolver.java [hunk 2]
```diff
@@ -174,11 +159,120 @@ private PartitionedMapTable createTable(
...
+        InternalIndex[] indexes = mapContainer.getIndexes().getIndexes();
+
+        if (indexes == null || indexes.length == 0) {
+            // Cannot resolve fields when the map doesn't have concurrent indexes
+            return FieldsMetadata.HD_ERROR;
+        }
+
+        InternalIndex index = indexes[0];
+
+        Iterator<QueryableEntry> entryIterator = index.getSqlRecordIterator();
```
Comment (petrpleshachkov) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/PartitionedMapTableResolver.java:202]: If we have `NULL` values in the index, the iterator will return an entry corresponding to it. I guess this is OK for further resolution logic?

**Review by viliam-durina — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/PartitionedMapTableResolver.java (hunk 2 above)
Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/PartitionedMapTableResolver.java:202]: We do not read index keys here, only actual entries. Therefore, the resolution logic is not affected.

**Review by petrpleshachkov — APPROVED** [review]

**Pull request merged**